### ISSUE TITLE
PHP error log

### ISSUE THREAD

**Issue by mikeruss1** [CIDRAM/CIDRAM#84]
got the following in the php error logfile

[02-Aug-2018 00:09:46 Europe/London] PHP Warning:  unlink(/home/xxxxxxx/public_html/cidram/vault/logfile20180702.log): No such file or directory in /home/xxxxx/public_html/cidram/vault/functions.php on line 1640

this was the first block on that day, and there are 3 duplicate blocks all timed 00:09:44

**Comment by Maikuolan** [comment]
Looked at the code at that part of the codebase. What's happened here, I believe, is that multiple instances of log rotation happened at the same time, due to race conditions (i.e., multiple requests came in at the same time, which each triggered log rotation; as one instance identified a file to be deleted, the other instance got to it first, causing this error to appear as a result of the former attempting to delete a file after the latter had already deleted it).

Similar situation to how log entries will occasionally have the same log entry ID if multiple block events occur at the same time, and not much I can really do about it unfortunately, beyond a complete rewrite of CIDRAM to not rely on flatfiles for logging and instead use a database solution like SQL or similar (which, I'd actually considered eventually doing for a future CIDRAM v2.0.0, at least as an optional alternative, but isn't on the scope of things to be done in the immediate future, at least).

It shouldn't hinder CIDRAM's ability to function normally at all, and won't cause any problems other than the annoyance of it occasionally popping up in error logs.

My feeling is that I'll probably mark this as a "wontfix" (at least, until a future time that flatfiles aren't necessary anymore), but I'll leave the issue open for a while, in case anyone else has any ideas they can share.

**Comment by mikeruss1** [comment]
indeed - all 3 blocks had the same ID

no problem, never happened before anyway

maybe use memcached to log events that are in progress ? Should be relatively easy ?

**Comment by Maikuolan** [comment]
> maybe use memcached to log events that are in progress ? Should be relatively easy ?

Not something I'm super familiar with (logging with memcached), so not sure how viable that is offhandedly. I'll look into that though.

**Comment by mikeruss1** [comment]
its basically a data base in memory. You store stuff with a key and obviously retrieval is very fast. So you can use it like session storage, but available to all users. It doesnt survive a restart so you need to check its there before using. And anything you store may need to be restored. Perfect for short term variables.

**Comment by mikeruss1** [comment]
not suggesting you log, the idea is you would know an event was going on, and could wait for completion.

**Comment by DanielRuf** [comment]
> not suggesting you log, the idea is you would know an event was going on, and could wait for completion.

In general we may have to lock files in this case or create some lockfile.

**Comment by DanielRuf** [comment]
But generally I would suggest using monolog.

**Comment by Dibbyo456** [comment]
@Maikuolan Is there any way I can contact you personally? Not regarding CIDRAM but need help with a logic. Email or chat or anything..?

**Comment by DanielRuf** [comment]
> @Maikuolan Is there any way I can contact you personally? Not regarding CIDRAM but need help with a logic. Email or chat or anything..?

Try https://gitter.im/Maikuolan

**Comment by Maikuolan** [comment]
I could probably help keep error logs cleaner by writing a small, custom error handler for CIDRAM, either to suppress errors like these entirely, or to just log them elsewhere, so that it won't clog up the default error logs too much.

It wouldn't address the actual cause of the problem, but it might alleviate some of the more annoying symptoms of it. There's also a possible small concern about error messages being exposed to the end-user being blocked, which might be undesirable in many cases.

Thoughts on this?

**Comment by DanielRuf** [comment]
If, then please use monolog (let's not reinvent the wheel).

**Comment by Maikuolan** [comment]
I hadn't replied anything else at this issue since the past year, due to that I hadn't yet figured out the best way to move forward, in terms of the "correct" way to log errors, the best way to allow Monolog integration, etc. Replying now, because I think I've figured something out and have began working on it today. I'll reply again, hopefully soon, after I've gotten a bit further with it, have tested, and confirmed it to be viable and working as intended.

**Comment by Maikuolan** [comment]
So.. Been talking around, with a few different people. Some have suggested the RFC 5424 approach, while others have suggested that the RFC 5424 approach would be completely overkill for something as simple as being able to log really simple, low-level errors/notices/warning/etc. Quite a few have suggested Monolog, but while some have suggested that I must implement it myself, others have suggested that I absolutely must not, and instead must allow other developers to implement it themselves, or decide for themselves whether to implement it, or to instead implement one of several other alternatives also suggested as things that could or should be implemented (there's also the question about the various log handlers supported by Monolog, and if I was to implement it myself, whether I would need to start then mirroring everything they support, which could result in very tight coupling to Monolog, and we don't want everything to be tightly coupled and so on). There's also been the suggestion to just write a simple logger myself, similar to as has already been done for logging other things, like block events (the "killed log"), reCAPTCHA events and so on, but disagreement regarding what it should look like exactly.. which then cycles back to the discussion about Monolog et al.

My solution (which I've committed just now): Adding a simple events orchestrator to CIDRAM (which I've also added to phpMussel, too), effectively providing a way for users to extend (or completely replace) parts of CIDRAM and its codebase without needing to actually touch/change/modify its codebase at all.

This means that users are now able to easily integrate Monolog into CIDRAM, to have Monolog act as either an additional pathway or replacement pathway for CIDRAM to send all its various logging tasks to, whether related to error logging, block event logging or whatever else. It also means that there doesn't need to be additional dependencies, tight couplings and so on for those that don't actually want it at all, means that I don't need to worrying about syncing up supported log handlers and whatever else, and means that I can keep the "you must support Monolog, but absolutely must not implement it yourself" crowd happy (in theory.. and time will tell, I guess).

For the "I want an out-of-the-box solution and I don't want to touch a single line of PHP code, ever" crowd: I'll likely get around to writing up a simple log handler of my own into CIDRAM at some point for logging errors, with relevant configuration directives and all the rest, thus providing an out-of-the-box solution to fall back onto, but I haven't quite gotten around to doing that yet (plus the aforementioned "disagreement regarding what it should look like exactly" problem still exists when it comes to that, so I'm not likely to rush into it, at least, not until I've given it however much more thought would be required to resolve that problem).

Anyway.. For those interested in integrating Monolog into CIDRAM (by way of the newly implemented events orchestrator), all you'll need to do (and I'll likely add this to the documentation at some point, so that users won't need to keep coming back to this particular issue comment in the future, but I haven't gotten around to doing that yet either; one step at a time), is create a new, empty PHP file, write a Monolog instantiation with any Monolog handlers you want to use, write whatever logging process you want to write (any data which needs to be logged should exist somewhere within the main CIDRAM array at the time the event is fired, meaning that it should be easy to leverage and work with), and put it all inside a call to CIDRAM's event orchestrator's `addHandler` method.

Example:

```PHP
<?php
use Monolog\Logger;
use Monolog\Handler\StreamHandler;

// Create handler for logging at CIDRAM side.
$CIDRAM['Events']->addHandler('writeToLog', function () {

    // New Monolog instance.
    $logger = new Logger('my_logger');

    // New Monolog handler.
    $logger->pushHandler(new StreamHandler(__DIR__.'/my_app.log', Logger::DEBUG));
    
    // Push CIDRAM log event to Monolog instance.
    $logger->info($Data);

});
```

(Example adapted from Monolog's own given examples, to be able to work as a CIDRAM event handler. Not tested, and would obviously need to be modified per the needs of whatever needs to be logged. Just shown here as an example only).

Save the new PHP file somewhere in CIDRAM's vault, cite it as a module at CIDRAM's module configuration directive (the earlier in the execution chain the better, in case of needing to log something caused by other modules), and you're done. CIDRAM should now be integrated with Monolog. :-)

For writing your own logging process, the logging event handlers that already exist in CIDRAM could be used as further examples and adapted for whatever you need (currently found in `functions.php`, but I've been thinking about migrating them out to their own file, so that may change in the future; regardless.. also included below):

```PHP

/**
 * Writes to the standard logfile.
 *
 * @return bool True on success; False on failure.
 */
$CIDRAM['Events']->addHandler('writeToLog', function () use (&$CIDRAM): bool {
    if (!$CIDRAM['Config']['general']['logfile'] || empty($CIDRAM['LogFileNames']['logfile'])) {
        return false;
    }
    $Data = !file_exists($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile']) || (
        $CIDRAM['Config']['general']['truncate'] > 0 &&
        filesize($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile']) >= $CIDRAM['ReadBytes']($CIDRAM['Config']['general']['truncate'])
    ) ? "\x3c\x3fphp die; \x3f\x3e\n\n" : '';
    $WriteMode = !empty($Data) ? 'w' : 'a';
    $Data .= $CIDRAM['ParseVars']($CIDRAM['Parsables'], $CIDRAM['FieldTemplates']['Logs'] . "\n");
    if ($CIDRAM['BuildLogPath']($CIDRAM['LogFileNames']['logfile'])) {
        $File = fopen($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile'], $WriteMode);
        fwrite($File, $Data);
        fclose($File);
        if ($WriteMode === 'w') {
            $CIDRAM['LogRotation']($CIDRAM['Config']['general']['logfile']);
        }
        return true;
    }
    return false;
});

/**
 * Writes to the Apache-style logfile.
 *
 * @return bool True on success; False on failure.
 */
$CIDRAM['Events']->addHandler('writeToLog', function () use (&$CIDRAM): bool {
    if (
        !$CIDRAM['Config']['general']['logfile_apache'] ||
        empty($CIDRAM['LogFileNames']['logfile_apache']) ||
        empty($CIDRAM['BlockInfo'])
    ) {
        return false;
    }
    $Data = sprintf(
        "%s - - [%s] \"%s %s %s\" %s %s \"%s\" \"%s\"\n",
        $CIDRAM['BlockInfo']['IPAddr'],
        $CIDRAM['BlockInfo']['DateTime'],
        $_SERVER['REQUEST_METHOD'] ?? 'UNKNOWN',
        $_SERVER['REQUEST_URI'] ?? '/',
        $_SERVER['SERVER_PROTOCOL'] ?? 'UNKNOWN/x.x',
        $CIDRAM['errCode'],
        strlen($CIDRAM['HTML']),
        $CIDRAM['BlockInfo']['Referrer'] ?? '-',
        $CIDRAM['BlockInfo']['UA'] ?? '-'
    );
    $WriteMode = !file_exists($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_apache']) || (
        $CIDRAM['Config']['general']['truncate'] > 0 &&
        filesize($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_apache']) >= $CIDRAM['ReadBytes']($CIDRAM['Config']['general']['truncate'])
    ) ? 'w' : 'a';
    if ($CIDRAM['BuildLogPath']($CIDRAM['LogFileNames']['logfile_apache'])) {
        $File = fopen($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_apache'], $WriteMode);
        fwrite($File, $Data);
        fclose($File);
        if ($WriteMode === 'w') {
            $CIDRAM['LogRotation']($CIDRAM['Config']['general']['logfile_apache']);
        }
        return true;
    }
    return false;
});

/**
 * Writes to the serialised logfile.
 *
 * @return bool True on success; False on failure.
 */
$CIDRAM['Events']->addHandler('writeToLog', function () use (&$CIDRAM): bool {
    if (
        !$CIDRAM['Config']['general']['logfile_serialized'] ||
        empty($CIDRAM['LogFileNames']['logfile_serialized']) ||
        empty($CIDRAM['BlockInfo'])
    ) {
        return false;
    }
    $BlockInfo = $CIDRAM['BlockInfo'];
    unset($BlockInfo['EmailAddr'], $BlockInfo['UALC'], $BlockInfo['favicon']);
    /** Remove empty entries prior to serialising. */
    $BlockInfo = array_filter($BlockInfo, function ($Value) {
        return !(is_string($Value) && empty($Value));
    });
    $WriteMode = !file_exists($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_serialized']) || (
        $CIDRAM['Config']['general']['truncate'] > 0 &&
        filesize($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_serialized']) >= $CIDRAM['ReadBytes']($CIDRAM['Config']['general']['truncate'])
    ) ? 'w' : 'a';
    if ($CIDRAM['BuildLogPath']($CIDRAM['LogFileNames']['logfile_serialized'])) {
        $File = fopen($CIDRAM['Vault'] . $CIDRAM['LogFileNames']['logfile_serialized'], $WriteMode);
        fwrite($File, serialize($BlockInfo) . "\n");
        fclose($File);
        if ($WriteMode === 'w') {
            $CIDRAM['LogRotation']($CIDRAM['Config']['general']['logfile_serialized']);
        }
        return true;
    }
    return false;
});
```

The `writeToLog` event is fired by CIDRAM's output generator, when logging a blocked request.

Currently, the other events are:

`error`: Fired by CIDRAM's error handler (this is what you'll want to use instead of `writeToLog`, if you want to log CIDRAM errors using Monolog).
`reCaptchaLog`: Fired by by the reCAPTCHA module, when logging a reCAPTCHA event (successfully completed reCAPTCHA or failed reCAPTCHA attempts).
`writeToPHPMailerEventLog`: Fired only when PHPMailer is installed as a CIDRAM component, specifically when it sends an email (and at no other times).

I may add more events to the list in the future, but there's no rush, and it should be okay for our needs at the moment. More things can be added in the future, if/when needed.

**Comment by Maikuolan** [comment]
Forgot to mention before.. Per the new events orchestrator, there's `addHandler`, and there's also `addHandlerFinal`:

```PHP
public function addHandler(string $Event, callable $Handler, bool $Replace = false): bool
public function addHandlerFinal(string $Event, callable $Handler, bool $Replace = false): bool
```

To suppress/replace CIDRAM's own internal event handlers, you'll want to set `$Replace` to `true` (can leave it alone if wanting to use Monolog as an additional pathway, rather than as a replacement pathway).

To prevent further handlers being added to the same event by other modules and so on, `addHandlerFinal` can be used instead of `addHandler`.

Anyway.. @DanielRuf Curious to know your thoughts on it; whether this should be sufficient regarding Monolog integration. :-)

(PS, for any non-PHP people reading this, if my most recent replies to this issue are going right over your head, I apologise. Said replies are more intended for those that don't mind writing their own code).

**Comment by Maikuolan** [comment]
Default error handler added by e559b135a8a75f41b58bab305a8f6fb6eb4f06dd + 8d090badc0ee796c2a5435514a6c8585707f426b. Nothing left to do at this issue anymore. Closing.
